features/294-login-without-specifying-login-id-key/api.md

```diff
@@ -0,0 +1,104 @@
+For supporting login without specifying `loginID`, have two proposals:
+
+## Proposal 1 - auth gear find corresponding user object automatically
+
+Keeping most of current implementation, auth gear finds possible user object automatically. Consider following use case:
+
+```javascript
+skygear.auth.signup({
+  email: 'example@example.com', 
+  username: 'example'
+}, 'password');
+
+// new:
+skygear.auth.login('example@example.com', 'password')
+.then((user) => {
+  console.log('login successfully');
+});
+```
+
+
+### Possible Problem
+
+If user signup with incorrect `loginIDMetadataKeys`, it may cause user identify miss assign. Consider following scenario:
+
+```javascript
+
+// signup user A
+skygear.auth.signup({
+  email: 'example@example.com', 
+  username: 'example'
+}, 'password');
+
+
+// signup user B
+// B's username is A's email
+skygear.auth.signup({
+  username: 'example@example.com', 
+  email: 'example'
+}, 'password');
+
+// doesn't know which user to match
+// user A: {email: 'example@example.com'}
+// user B: {username: 'example@example.com'}
+skygear.auth.login('example@example.com', 'password');
+```
+
+### Code change
+
+Just brute force and find all possible principals. 
+
+When auth gear is got a login request without specifying `loginIDMetadataKey`, auth gear will try to find all possible principals based on current `loginIDMetadataKeys` setting. 
+
+For example, if a user login with string `"example"`, and `loginIDMetadataKeys=[["username"], ["email"], ["phone"]]`, auth principal store finds all principals which has `{username: "example"}`, `{email: "example"}` and `{phone: "example"}` in `authdata` field.
+
+And then login handler will check input password with stored password, if **only one** of the principals matched, then the user login successfully. If there are more than one principals matched, auth gear will return `Ambiguous loginID` error to the client. 
+
+The major drawback of this solution is the API UX is not good, basically just throw the problem back to the app developer. But the situation of two users uses same password with interchanged `loginID` should be a rare case.
+
+
+## Proposal 2 - meaningless login ID
```

Comment:
Actually, I like this idea more as I originally suggested something like this lol.

The beauty of it is:
1. We can remove a configuration variable: `LOGIN_ID_METADATA_KEYS`
1. We can simply have the idea of {loginID + password} pair to even support different passwords.

On the API, I suggest we should have a `loginID-type` to indicate the loginID name, so it could be identified in request/sessions.

```
skygear.auth.signup(loginID: String, password: String, loginID-type: String=null)
skygear.auth.login(loginID: String, password: String, loginID-type: String=null)
```

Beside sthat:
* We can still have `signupWithUsername/email`, or login, where `loginID-type` is set to `Username` and `Email` respectively
* We can still have `loginID` as Object instead of String, where we have a standard way to serialize it, but maybe developers don't need it?
* In user object, we need to have a dict of LoginIDs of the user / current users

Reply:
In the latest update, I do some adjustments based on your comments, please refer the file for the detail:

Summaries:
> `signup(loginID: String, password: String, loginID-type: String=null, data: Object=null)` 

is quite verbose (include data filed), and because we have to present the `loginID` in `metadata`, so the `loginID-type` can't be default to `null`. Thus I keep using `signup(loginIDs: dict, password: String, data: Object)`

> skygear.auth.login(loginID: String, password: String, loginID-type: String=null)

I think login no need specifying `loginID-type` here.

> In user object, we need to have a dict of LoginIDs of the user / current users

In [253-auth-sdk-callback-in-next](https://github.com/SkygearIO/features/blob/94f5dd8d22d0107c3e48591ab0cc5448b4b0eaf1/features/253-auth-sdk-callback-in-next/api.md), we have `loginID`s in `metadata` already.

> We can still have loginID as Object instead of String, where we have a standard way to serialize it, but maybe developers don't need it?

I added a section to show to show my preferred way to support loginID as Object.